Add Event Classes

tmcahill opened the pull request:
First of a series of updates for my implementation of Alisha Chromey's work for 4DMLM.

This update implements the `EventClass` object, a simple data structure by which event classes are implemented.

EventClasses are created when using event class mode (-ec). The user may provide one or more EAs per EA tag in the runlist to create an `EventClass` for each EA in the tag. Event classes are paired to the data files of the same ID. [Example runlist](https://veritas.sao.arizona.edu/wiki/V2dl3_dev_notes#Event_Classes)

Giving event classes their own object is cleaner than passing event class data many times via function parameters. It also allows us to avoid redundant computations for input files which correspond to an already constructed event class.

Next patch will make use of these EventClasses to select events to separate .fits files.

Vegas-side run successful with this patch.

## Itemized changes
EventClass.py 
- new file
- object stores event class data (e.g the effective area file and its relevant cuts)

genHDUList.py
- optional keyword arg `event_class` to pass `EventClass`es to `VegasDataSource`
- exceptions to ensure `event_class` xor `effective_area_file` is None

v2dl3.py
- `runlist_to_file_pair` extended to support `EventClass`
- `-ec` flag added to trigger event class mode
- cleaned redundant `loadROOTfiles()` call by assigning `file_type` appropriately

VegasDataSource.py
- `event_classes` optional keyword argument and class member
- exceptions to ensure `event_classes` xor `ea_file` is None

parseSt6RunList.py
- `validateRunlist()` extended to support multiple EAs per tag in event class mode

util.py
- `getCuts()` added; retrieves cuts values from the given keys in a ROOT cuts text

GernotMaier commented on pyV2DL3/EventClass.py:
```diff
@@ -0,0 +1,64 @@
+import ROOT
```
This is a VEGAS specific module, which requires access to data products like effective areas which are only available in this format for VEGAS. My suggestion would be to move this module into the pyV2Dl3/vegas directory. This way we have clearly separated modules required by both chains (in ./pyV2Dl3) and those required for one or the other.

GernotMaier reviewed (changes requested):
This looks like a very nice addition with the functionality of event classes.

Two comments:
- move vegas specific modules into the vegas directory
- Suggest to rename v2dl3.py to vdl3_for_vegas and remove all the eventdisplay specifics from this file. We are using already for Eventdisplay a different main script (v2dl3_for_eventdisplay). Given that only vegas has the eventclass functionaliy, my suggestion is to introduce a clear split now.

tmcahill commented:
Combed through every occurrence of 'v2dl3' in the code as part of the split for `v2dl3.py` -> `v2dl3_for_vegas.py`:
Renamed `environment.yml` -> `environment-vegas.yml`
Renamed `v2dl3` conda env to `v2dl3-vegas` conda env
Changed `v2dl3.py` to log help message upon being called with any flags (or lack thereof)
Added `v2dl3-vegas` entry point for `v2dl3_for_vegas.py`
Updated `README.md` to reflect above changes

Tested the updated README commands on a fresh install to a new conda env; All working and successfully processed a runlist.

Other minor changes are EventClass moved to vegas-side and a small CPyCppyy exception bugfix for vegas-side. See commits.

GernotMaier reviewed (approved):
All good - no comments from my side.
